Agent/Public/NewRelic.m

```diff
@@ -600,10 +601,63 @@ + (BOOL) incrementAttribute:(NSString*)name
                                                                                       persistent:YES];
 }
 
-+ (BOOL) setUserId:(NSString*)userId {
-    return [[NewRelicAgentInternal sharedInstance].analyticsController setSessionAttribute:@"userId"
-                                                                                     value:userId
-                                                                                persistent:YES];
++ (BOOL) setUserId:(NSString* _Nullable)userId {
+
+    /*
+     
+     1. When setUserID(value: string|null) is called:
+        a. If userID was previously null and new value is non-null:
+            i. continue the current session and set the new userID
+        b. If userID was previously not-null and new value is different (including null):
+            i. end the current session and perform harvest
+            ii. start a new session with the new userID
+     */
+    NSString *previousUserId = [[NewRelicAgentInternal sharedInstance] getUserId];
+    BOOL newSession = false;
+    // If the client passes a new userId that is non NULL.
+    if (userId != NULL) {
+        // If userId has not previously been set.
+        if (previousUserId == NULL) {
+            // continue session and set the new userID
+        }
+        // A new userId has been set where the previous one was not NULL.
+        else {
+            newSession = true;
+        }
+    }
+    // If the client passes a new NULL userId.
+    else {
+        if (previousUserId == NULL) {
+            // Do nothing if passed user id is null and saved userId is null.
+        }
+        else {
+            // end session and harvest.
+            newSession = true;
+        }
+    }
```

Comment:
It looks like you wrote the if statements to be deliberate to what was required in the spec but I don't like the empty conditions. Doesn't hurt anything but looks weird.

Reply:
Fixed.